src/core/execute.c

```diff
@@ -720,74 +720,129 @@ static int ask_for_confirmation(char *response, char **argv) {
         return r;
 }
 
-static int enforce_groups(const ExecContext *context, const char *username, gid_t gid) {
+static int get_fixed_user(const ExecContext *c, const char **user,
+                          uid_t *uid, gid_t *gid,
+                          const char **home, const char **shell) {
+        assert(c);
+
+        if (!c->user)
+                return 0;
+
+        /* Note that we don't set $HOME or $SHELL if they are not particularly enlightening anyway
+         * (i.e. are "/" or "/bin/nologin"). */
+
+        *user = c->user;
+        return get_user_creds_clean(user, uid, gid, home, shell);
+}
```

Comment:
we generally try to avoid functions that clobber their call-by-ref parameters on failure. In this case that's what happens if get_user_creds_clean() fails: *user is manipulated anyway.

Not that it would really matter in this case, but it's more a general conceptual choice...